File: app/src/main/java/com/bis5/fitjourney/adapters/ActiveSetAdapter.java
package com.bis5.fitjourney.adapters;

import android.os.CountDownTimer;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.bis5.fitjourney.R;
import com.bis5.fitjourney.models.Exercise;
import com.google.android.material.button.MaterialButton;
import com.google.android.material.textfield.TextInputLayout;

import java.util.HashSet;
import java.util.Locale;

public class ActiveSetAdapter extends RecyclerView.Adapter<ActiveSetAdapter.ActiveSetViewHolder> {

    public interface OnLogSetListener {
        void onLogSet(int setNumber, double weight, int reps);
    }

    private final Exercise exercise;
    private final OnLogSetListener onLogSetListener;
    private final HashSet<Integer> loggedSets = new HashSet<>();
    private CountDownTimer timer;
    private int restingSetPosition = -1;

    public ActiveSetAdapter(Exercise exercise, OnLogSetListener listener) {
        this.exercise = exercise;
        this.onLogSetListener = listener;
    }

    @NonNull
    @Override
    public ActiveSetViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_active_set, parent, false);
        return new ActiveSetViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull ActiveSetViewHolder holder, int position) {
        int setIndex = position + 1;
        holder.setNumber.setText(String.format(Locale.getDefault(), "Set %d", setIndex));

        if (restingSetPosition == position) {
            holder.weightLayout.setVisibility(View.GONE);
            holder.repsLayout.setVisibility(View.GONE);
            holder.logSetButton.setVisibility(View.GONE);
            holder.restTimer.setVisibility(View.VISIBLE);
            holder.skipRestButton.setVisibility(View.VISIBLE);

            holder.skipRestButton.setOnClickListener(v -> {
                if (timer != null) {
                    timer.cancel();
                }
                restingSetPosition = -1;
                notifyItemChanged(position);
            });
        } else if (loggedSets.contains(setIndex)) {
            holder.weightLayout.setVisibility(View.VISIBLE);
            holder.repsLayout.setVisibility(View.VISIBLE);
            holder.weight.setEnabled(false);
            holder.reps.setEnabled(false);
            holder.logSetButton.setClickable(false);
            holder.logSetButton.setVisibility(View.VISIBLE);
            // BUG FIX: Use setIconTintResource for MaterialButton
            holder.logSetButton.setIconTintResource(android.R.color.holo_green_dark);
            holder.restTimer.setVisibility(View.GONE);
            holder.skipRestButton.setVisibility(View.GONE);
        } else {
            holder.weightLayout.setVisibility(View.VISIBLE);
            holder.repsLayout.setVisibility(View.VISIBLE);
            holder.weight.setEnabled(true);
            holder.reps.setEnabled(true);
            holder.weight.setText(String.valueOf(exercise.getWeight()));
            holder.reps.setText(String.valueOf(exercise.getReps()));
            holder.logSetButton.setClickable(true);
            holder.logSetButton.setVisibility(View.VISIBLE);
            // BUG FIX: Use setIconTint(null) to clear tint for MaterialButton
            holder.logSetButton.setIconTint(null);
            holder.restTimer.setVisibility(View.GONE);
            holder.skipRestButton.setVisibility(View.GONE);

            holder.logSetButton.setOnClickListener(v -> {
                String weightStr = holder.weight.getText().toString();
                String repsStr = holder.reps.getText().toString();
                double weightValue = weightStr.isEmpty() ? 0.0 : Double.parseDouble(weightStr);
                int repsValue = repsStr.isEmpty() ? 0 : Integer.parseInt(repsStr);
                
                onLogSetListener.onLogSet(setIndex, weightValue, repsValue);
                loggedSets.add(setIndex);
                startRestTimer(holder, position);
            });
        }
    }

    private void startRestTimer(ActiveSetViewHolder holder, int position) {
        restingSetPosition = position;
        notifyItemChanged(position);

        timer = new CountDownTimer(60000, 1000) { // 60-second timer
            @Override
            public void onTick(long millisUntilFinished) {
                long seconds = millisUntilFinished / 1000;
                holder.restTimer.setText(String.format(Locale.getDefault(), "%02d:%02d", seconds / 60, seconds % 60));
            }

            @Override
            public void onFinish() {
                restingSetPosition = -1;
                notifyItemChanged(position);
            }
        }.start();
    }

    @Override
    public int getItemCount() {
        return exercise.getSets();
    }

    @Override
    public void onDetachedFromRecyclerView(@NonNull RecyclerView recyclerView) {
        super.onDetachedFromRecyclerView(recyclerView);
        if (timer != null) {
            timer.cancel(); // Prevent memory leaks
        }
    }

    public static class ActiveSetViewHolder extends RecyclerView.ViewHolder {
        final TextView setNumber;
        final TextInputLayout weightLayout;
        final TextInputLayout repsLayout;
        final EditText weight;
        final EditText reps;
        // BUG FIX: This is a MaterialButton, not an ImageButton
        final MaterialButton logSetButton;
        final TextView restTimer;
        final Button skipRestButton;

        public ActiveSetViewHolder(@NonNull View itemView) {
            super(itemView);
            setNumber = itemView.findViewById(R.id.tvSetNumber);
            weightLayout = itemView.findViewById(R.id.tilWeight);
            repsLayout = itemView.findViewById(R.id.tilReps);
            weight = itemView.findViewById(R.id.etWeight);
            reps = itemView.findViewById(R.id.etReps);
            logSetButton = itemView.findViewById(R.id.btnLogSet);
            restTimer = itemView.findViewById(R.id.tvRestTimer);
            skipRestButton = itemView.findViewById(R.id.btnSkipRest);
        }
    }
}

File: app/src/main/java/com/bis5/fitjourney/adapters/ActiveSetAdapter.kt
package com.bis5.fitjourney.adapters

import android.os.CountDownTimer
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.EditText
import android.widget.ImageButton
import android.widget.TextView
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.RecyclerView
import com.bis5.fitjourney.R
import com.bis5.fitjourney.models.Exercise
import com.google.android.material.textfield.TextInputLayout

class ActiveSetAdapter(
    private val exercise: Exercise,
    private val onLogSet: (setNumber: Int, weight: Double, reps: Int) -> Unit
) : RecyclerView.Adapter<ActiveSetAdapter.ActiveSetViewHolder>() {

    private val loggedSets = mutableSetOf<Int>()
    private var timer: CountDownTimer? = null
    private var restingSetPosition: Int = -1

    inner class ActiveSetViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val setNumber: TextView = itemView.findViewById(R.id.tvSetNumber)
        val weightLayout: TextInputLayout = itemView.findViewById(R.id.tilWeight)
        val repsLayout: TextInputLayout = itemView.findViewById(R.id.tilReps)
        val weight: EditText = itemView.findViewById(R.id.etWeight)
        val reps: EditText = itemView.findViewById(R.id.etReps)
        val logSetButton: ImageButton = itemView.findViewById(R.id.btnLogSet)
        val restTimer: TextView = itemView.findViewById(R.id.tvRestTimer)
        val skipRestButton: Button = itemView.findViewById(R.id.btnSkipRest)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ActiveSetViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.item_active_set, parent, false)
        return ActiveSetViewHolder(view)
    }

    override fun onBindViewHolder(holder: ActiveSetViewHolder, position: Int) {
        val setIndex = position + 1
        holder.setNumber.text = "Set $setIndex"

        when {
            restingSetPosition == position -> {
                // This set is currently in a rest period
                holder.weightLayout.visibility = View.GONE
                holder.repsLayout.visibility = View.GONE
                holder.logSetButton.visibility = View.GONE
                holder.restTimer.visibility = View.VISIBLE
                holder.skipRestButton.visibility = View.VISIBLE

                holder.skipRestButton.setOnClickListener {
                    timer?.cancel()
                    restingSetPosition = -1
                    notifyItemChanged(position)
                }
            }
            loggedSets.contains(setIndex) -> {
                // This set has been logged
                holder.weightLayout.visibility = View.VISIBLE
                holder.repsLayout.visibility = View.VISIBLE
                holder.weight.isEnabled = false
                holder.reps.isEnabled = false
                holder.logSetButton.isClickable = false
                holder.logSetButton.visibility = View.VISIBLE
                holder.logSetButton.setColorFilter(ContextCompat.getColor(holder.itemView.context, android.R.color.holo_green_dark))
                holder.restTimer.visibility = View.GONE
                holder.skipRestButton.visibility = View.GONE
            }
            else -> {
                // This set is ready to be logged
                holder.weightLayout.visibility = View.VISIBLE
                holder.repsLayout.visibility = View.VISIBLE
                holder.weight.isEnabled = true
                holder.reps.isEnabled = true
                holder.weight.setText(exercise.weight.toString())
                holder.reps.setText(exercise.reps.toString())
                holder.logSetButton.isClickable = true
                holder.logSetButton.visibility = View.VISIBLE
                holder.logSetButton.clearColorFilter()
                holder.restTimer.visibility = View.GONE
                holder.skipRestButton.visibility = View.GONE

                holder.logSetButton.setOnClickListener {
                    val weightValue = holder.weight.text.toString().toDoubleOrNull() ?: 0.0
                    val repsValue = holder.reps.text.toString().toIntOrNull() ?: 0
                    onLogSet(setIndex, weightValue, repsValue)

                    loggedSets.add(setIndex)
                    startRestTimer(holder, position)
                }
            }
        }
    }

    private fun startRestTimer(holder: ActiveSetViewHolder, position: Int) {
        restingSetPosition = position
        notifyItemChanged(position) // Update UI to show timer

        timer = object : CountDownTimer(60000, 1000) { // 60-second timer
            override fun onTick(millisUntilFinished: Long) {
                val seconds = millisUntilFinished / 1000
                holder.restTimer.text = String.format("%02d:%02d", seconds / 60, seconds % 60)
            }

            override fun onFinish() {
                restingSetPosition = -1
                notifyItemChanged(position) // Update UI to hide timer
            }
        }.start()
    }

    override fun getItemCount(): Int {
        return exercise.sets
    }

    override fun onDetachedFromRecyclerView(recyclerView: RecyclerView) {
        super.onDetachedFromRecyclerView(recyclerView)
        timer?.cancel() // Prevent memory leaks
    }
}